### DISCUSSION TITLE
Organising rules in the manager service

### DISCUSSION THREAD

**Discussion by jonathonherbert** [guardian/typerighter#313] [Ideas]
In our google sheet, there are two cells we can add metadata to help us organise rules – the **category**, a single, mandatory value, and the **tags**, an optional comma-separated list of strings. 

Examples for **categories** include ‘Style guide’, ‘Check this’, ‘Names’, ‘Style guide and names’, ‘Typos’, ‘Dates’,’Style guide and names’, ‘Typography’. Examples for **tags** include ‘Names’, ‘MP’, ‘General’, ‘SG’ (Style guide), ‘Tokyo2020’ (since removed!).

We can see a few problems here:

* There’s repetition – both within and across categories and tags.
* It’s not clear what each sort of entity is for. ‘Style Guide’ is repeated in both categories and tags – where should it live?

In figuring out how to proceed, it’s worth clarifying the question – what user stories are we attempting to satisfy with categories and tags? Do we need either, both, or something else?

It’s worth noting that while there’s plenty of overlap between tags and categories as they stand, we can see some legitimate use cases for more than one category emerge:

* Marking a rule that relates to offensive language - 2 uses
* Marking a rule that relates to a legal dispute - 1 use
* Adding detail to a rule – for example, a rule with category **Name** that is tagged as an **MP** - 653 uses
* Regional differences, e.g. **US Spelling. ** (Are tags the right vehicle for this? Eventually, hopefully not – but we may be without a dictionary, let alone a US-specific dictionary, for some time.) - 6 uses (but 27 more that reference US Spelling, and should probably have this tag)

### 1. As a rule user, I can see a high-level indication of where my rule has come from.

At the moment we use categories to do this. The uppercase text at the top of the match tooltip shows the category:

![Screenshot 2023-04-19 at 08 56 06](https://user-images.githubusercontent.com/7767575/233659431-851c16b9-39ec-4fc3-b154-95a2bdd7f1aa.png)

‘Style guide and names’ is obviously not specific enough, but the hypothesis is that it’s useful to know where a rule is coming from. This is related to Typerighter’s vision of [transparency](https://github.com/guardian/typerighter/blob/main/vision.md#transparent) – ‘When Typerighter matches text, it will be clear why that match has occurred.’

### 2. As a rule owner, I can identify rules that relate to one another, making it possible to discover all the rules that belong to a particular domain – and remove them.

At the moment rule owners use categories to do this. One example above is ‘Tokyo 2020’. The rules for Tokyo are now gone – no longer being relevant – and each rule having the same category made it easier to find them all and delete them once they’d served their purpose.

Are there more stories we can think of?

### A few options:

1. **Just categories** – a single, mandatory relationship between a rule and a ‘category’ that describes what sort of rule it is.
    * ✅ Simple.
    * ✅ Satisfies story 1 & 2 separately.
    * ✅ Good enough – we could keep the ‘tags’ field as plain text, and introduce structure later as necessary.
    * ❌ Not yet able to describe multiple attributes in a structured way.
2. **Just tags** – the ability to optionally assign zero-to-many tags to a rule
    * ✅ More flexible – we can add more than one tag to a rule.
    * ❓ How do we display tags to the user (per the screenshot above?) Does the first tag display? Do all tags display? Are some tags ‘presentational’, while others remain hidden?
    * ❓ Do we want to support untagged rules? Perhaps a rule must have a tag to be considered publishable.
    * ❌ More complicated (although not much more complicated) to implement and maintain.
3. **Both**
    * ✅  Very flexible – we can add more categories to help describe our corpus more flexibly.
    * ✅  Arguably easier to solve the problem of what to present to the rule user, as categories are already used for this purpose and work well.
    * ❓ Is it clear what these different entities are for to users? If not, we could end up with odd or overlapping sets of categories/tags.
    * ❌ More complicated to implement and maintain.

## Appendix: our current categories and tags

<details>
**Categories:**

<table>
  <tr>
   <td>Check this
   </td>
  </tr>
  <tr>
   <td>Guardian convention
   </td>
  </tr>
  <tr>
   <td>Style guide and names
   </td>
  </tr>
  <tr>
   <td>General
   </td>
  </tr>
  <tr>
   <td>Names
   </td>
  </tr>
  <tr>
   <td>Typos
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: General
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Diving
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Equestrian
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Cycling
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Rugby
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Boxing
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Football
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Swimming
   </td>
  </tr>
  <tr>
   <td>Coronavirus
   </td>
  </tr>
  <tr>
   <td>Typography
   </td>
  </tr>
  <tr>
   <td>Dates
   </td>
  </tr>
  <tr>
   <td>Style Guide
   </td>
  </tr>
</table>

**Tags**

<table>
  <tr>
   <td>SG
   </td>
  </tr>
  <tr>
   <td>General
   </td>
  </tr>
  <tr>
   <td>US spelling
   </td>
  </tr>
  <tr>
   <td>Unnecessary contraction?
   </td>
  </tr>
  <tr>
   <td>Swearword!
   </td>
  </tr>
  <tr>
   <td>13?
   </td>
  </tr>
  <tr>
   <td>Offensive?
   </td>
  </tr>
  <tr>
   <td>US term
   </td>
  </tr>
  <tr>
   <td>Taste?
   </td>
  </tr>
  <tr>
   <td>Names
   </td>
  </tr>
  <tr>
   <td>Coronavirus
   </td>
  </tr>
  <tr>
   <td>Poss typo
   </td>
  </tr>
  <tr>
   <td>SG
   </td>
  </tr>
  <tr>
   <td>Guardian convention
   </td>
  </tr>
  <tr>
   <td>Legal
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020
   </td>
  </tr>
  <tr>
   <td>Names, MP
   </td>
  </tr>
  <tr>
   <td>SG, LanguageTool, Semantics
   </td>
  </tr>
  <tr>
   <td>LanguageTool
   </td>
  </tr>
  <tr>
   <td>SG, LanguageTool, dates
   </td>
  </tr>
  <tr>
   <td>Typography, LanguageTool
   </td>
  </tr>
  <tr>
   <td>SG, LanguageTool, Typography
   </td>
  </tr>
  <tr>
   <td>SG, LanguageTool
   </td>
  </tr>
  <tr>
   <td>Typos, LanguageTool
   </td>
  </tr>
</table>

# Organising rules in Typerighter

In our google sheet, there are two cells we can add metadata to help us organise rules – the **category**, a single, mandatory value, and the **tags**, an optional comma-separated list of strings. 

Examples for **categories** include ‘Style guide’, ‘Check this’, ‘Names’, ‘Style guide and names’, ‘Typos’, ‘Dates’,’Style guide and names’, ‘Typography’. Examples for **tags** include ‘Names’, ‘MP’, ‘General’, ‘SG’ (Style guide), ‘Tokyo2020’ (since removed!).

We can see a few problems here:

* There’s repetition – both within and across categories and tags.
* It’s not clear what each sort of entity is for. ‘Style Guide’ is repeated in both categories and tags – where should it live?

In figuring out how to proceed, it’s worth clarifying the question – what user stories are we attempting to satisfy with categories and tags? Do we need either, both, or something else?

It’s worth noting that while there’s plenty of overlap between tags and categories as they stand, we can see some legitimate use cases for more than one category emerge:

* Marking a rule that relates to offensive language - 2 uses
* Marking a rule that relates to a legal dispute - 1 use
* Adding detail to a rule – for example, a rule with category **Name** that is tagged as an **MP - **653 uses
* Regional differences, e.g. **US Spelling. **(Are tags the right vehicle for this? Eventually, hopefully not – but we may be without a dictionary, let alone a US-specific dictionary, for some time.) - 6 uses (but 27 more that reference US Spelling, and should probably have this tag)
1. As a rule user, I can see a high-level indication of where my rule has come from.

At the moment we use categories to do this. The uppercase text at the top of the match tooltip shows the category:

<p id="gdcalert1" ><span style="color: red; font-weight: bold">>>>>>  gd2md-html alert: inline image link here (to images/image1.png). Store image on your image server and adjust path/filename/extension if necessary. </span><br>(<a href="#">Back to top</a>)(<a href="#gdcalert2">Next alert</a>)<br><span style="color: red; font-weight: bold">>>>>> </span></p>

![alt_text](images/image1.png "image_tooltip")

‘Style guide and names’ is obviously not specific enough, but the hypothesis is that it’s useful to know where a rule is coming from. This is related to Typerighter’s vision of [transparency](https://github.com/guardian/typerighter/blob/main/vision.md#transparent) – ‘When Typerighter matches text, it will be clear why that match has occurred.’

2. As a rule owner, I can identify rules that relate to one another, making it possible to discover all the rules that belong to a particular domain – and remove them.

At the moment rule owners use categories to do this. One example above is ‘Tokyo 2020’. The rules for Tokyo are now gone – no longer being relevant – and each rule having the same category made it easier to find them all and delete them once they’d served their purpose.

Are there more stories we can think of?

### A few options:

1. **Just categories** – a single, mandatory relationship between a rule and a ‘category’ that describes what sort of rule it is.
    * ✅Simple.
    * ✅Satisfies story 1 & 2 separately.
    * ✅Good enough – we could keep the ‘tags’ field as plain text, and introduce structure later as necessary.
    * ❌Not yet able to describe multiple attributes in a structured way.
2. **Just tags** – the ability to optionally assign zero-to-many tags to a rule
    * ✅More flexible – we can add more than one tag to a rule.
    * ❓How do we display tags to the user (per the screenshot above?) Does the first tag display? Do all tags display? Are some tags ‘presentational’, while others remain hidden?
    * ❓Do we want to support untagged rules? Perhaps a rule must have a tag to be considered publishable.
    * ❌More complicated (although not much more complicated) to implement and maintain.
3. **Both**
    * ✅ Very flexible – we can add more categories to help describe our corpus more flexibly.
    * ✅ Arguably easier to solve the problem of what to present to the rule user, as categories are already used for this purpose and work well.
    * ❓Is it clear what these different entities are for to users? If not, we could end up with odd or overlapping sets of categories/tags.
    * ❌More complicated to implement and maintain.

## Appendix: our current categories and tags

Categories:

<table>
  <tr>
   <td>Check this
   </td>
  </tr>
  <tr>
   <td>Guardian convention
   </td>
  </tr>
  <tr>
   <td>Style guide and names
   </td>
  </tr>
  <tr>
   <td>General
   </td>
  </tr>
  <tr>
   <td>Names
   </td>
  </tr>
  <tr>
   <td>Typos
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: General
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Diving
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Equestrian
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Cycling
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Rugby
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Boxing
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Football
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020: Swimming
   </td>
  </tr>
  <tr>
   <td>Coronavirus
   </td>
  </tr>
  <tr>
   <td>Typography
   </td>
  </tr>
  <tr>
   <td>Dates
   </td>
  </tr>
  <tr>
   <td>Style Guide
   </td>
  </tr>
</table>

Tags:

<table>
  <tr>
   <td>SG
   </td>
  </tr>
  <tr>
   <td>General
   </td>
  </tr>
  <tr>
   <td>US spelling
   </td>
  </tr>
  <tr>
   <td>Unnecessary contraction?
   </td>
  </tr>
  <tr>
   <td>Swearword!
   </td>
  </tr>
  <tr>
   <td>13?
   </td>
  </tr>
  <tr>
   <td>Offensive?
   </td>
  </tr>
  <tr>
   <td>US term
   </td>
  </tr>
  <tr>
   <td>Taste?
   </td>
  </tr>
  <tr>
   <td>Names
   </td>
  </tr>
  <tr>
   <td>Coronavirus
   </td>
  </tr>
  <tr>
   <td>Poss typo
   </td>
  </tr>
  <tr>
   <td>SG
   </td>
  </tr>
  <tr>
   <td>Guardian convention
   </td>
  </tr>
  <tr>
   <td>Legal
   </td>
  </tr>
  <tr>
   <td>Tokyo 2020
   </td>
  </tr>
  <tr>
   <td>Names, MP
   </td>
  </tr>
  <tr>
   <td>SG, LanguageTool, Semantics
   </td>
  </tr>
  <tr>
   <td>LanguageTool
   </td>
  </tr>
  <tr>
   <td>SG, LanguageTool, dates
   </td>
  </tr>
  <tr>
   <td>Typography, LanguageTool
   </td>
  </tr>
  <tr>
   <td>SG, LanguageTool, Typography
   </td>
  </tr>
  <tr>
   <td>SG, LanguageTool
   </td>
  </tr>
  <tr>
   <td>Typos, LanguageTool
   </td>
  </tr>
</table>

**Comment by pradasa** [comment]
This is brilliant. Thanks for making sense of this, Jon!

I can’t think of more users right now and looking at options, I’d be inclined to go with **3.Both** as long as it’s very clear to rule owners what is a tag vs. what is a category.

If we take a step back from the idea of categories and tags and look at everything we have, we see that we are using terms that relate to:

What is the **origin** of a rule
- Styleguide and names, SG, Style guide (there are 3 separate items for this) -> Is this in relation to what House Style designates as the style guide?
- Guardian convention
- Legal

What is the **purpose** of a rule
- Check this (worth discussing more what this means. Is this duplicating actionable/ ambiguous?)

What is the **content** of a rule
- Names
- Typos
- Dates
- Typography
- US spelling
- Swearword
- Etc.

What **topic/event/story** does the rule relate to?
- Tokyo 2020
- MPs
- Coronavirus

It may be worth looking at the language to make it clearer to rule owners.
If we find a way to rationalise the above, we might get more benefits for using categories + tags. 

One thing to consider is if this rationalisation would entail having to restructure the content of the existing rules and how would we go about it so we don’t have to go rule by rule to do so.

**Comment by jonathonherbert** [comment]
@pradasa really useful, so in your list we've got

**origin, purpose, content, topic**

- A rule will always have an **origin**. Perhaps that's a good candidate for a one-one relationship, like our 'category' property.
- I think the **purpose** will always be related to the rule colour, which is automatically derived ('I'd like you to know this is correct.', 'This matches the style guide and it's a candidate for checking', 'I think this needs a correction'), so although it's a concept that's useful, I don't think it belongs in the taxonomy.
- I like **content** – specifically, I guess this is the content the rule is targeting? Maybe **target content**? Maybe **content** is unambiguous enough?

So perhaps

- **origin** – one-to-one relationship with rule, like our 'category' as it stands
- **content**, **topic** – types of tags

What do you think?

**Reply by pradasa** [reply]
Sorry for the late reply, I just saw your comment on chat. 
I think reducing to origin and tags makes sense!
